xonsh/environ.py

```diff
@@ -1985,20 +1985,39 @@ def __getitem__(self, key):
         ):
             self._detyped = None
 
-        validator = self.get_validator(key)
-        converter = self.get_converter(key)
-        if not validator(val):
-            val = converter(val)
         return val
 
     def __setitem__(self, key, val):
         validator = self.get_validator(key)
         converter = self.get_converter(key)
         detyper = self.get_detyper(key)
+        old_value = self._d[key] if key in self._d else self._no_value
         if not validator(val):
-            val = converter(val)
+            old_value_validate = validator(old_value)
+            try:
+                val = converter(val)
```

Comment:
I pretty strongly feel that any errors raised by the conveters should be passed up to the user.

Reply:
Let's discuss about the real life example. There is `$AUTOJUMP_PRINT_DIR` from `autojump` package.  This variable has `False` value.
I want to register `*_DIR` as `path` variables. Should I get notifications about error of converting `$AUTOJUMP_PRINT_DIR` to `path` every day and every time I run the shell? 
My answer is no. I suggest convert variables which we can convert and leave unconverted others.
If I'll face with `$AUTOJUMP_PRINT_DIR` should be `path` but I see `bool` I'll register it if I want.

Reply:
Don't forget that we can register the variable any time. For example you have 10 xontribs with different `*_DIR` variables with different types. Along your suggestion we should show i.e. 5 error messages to user about wrong convertation. Right? It's very annoying for user because it don't want to resolve this. In the current logic we just convert what we can and leave other values in original value - it is non breaking logic.